cpp/src/arrow/compute/kernel.cc

```diff
@@ -411,21 +409,23 @@ bool InputType::Matches(const DataType& type) const {
       return type_->Equals(type);
     case InputType::USE_TYPE_MATCHER:
       return type_matcher_->Matches(type);
-    default:
-      // ANY_TYPE
-      return true;
+    case InputType::ANY_TYPE:
+      break;
   }
+  return true;
```

Comment:
This means that if `kind_` is poisoned memory, we return `true` instead of poison (true or false non-deterministically). Making the code identical to the previous version but without the `default` case.